File: week08/19/P5.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    int findMinHelp(vector<int> nums, int start, int end) {
        if (end - start == 0) return nums[start];
        int mid = (start + end) / 2;
        int left = findMinHelp(nums, start, mid);
        int right = findMinHelp(nums, mid+1, end);
        return min(left, right);
    }

    int findMin(vector<int> nums) {
        return findMinHelp(nums, 0, nums.size()-1);
    }
};

int main() {
    // vector<int> nums = {1,3,5};
    vector<int> nums = {2,2,2,0,1};
    Solution result;
    cout << result.findMin(nums) << endl;

    return 0;
}

File: week08/19/P5.py
class Solution(object):
    def findMin(self, nums):
        return self.findMinHelp(nums, 0, len(nums)-1)
    
    def findMinHelp(self, nums, start, end):
        if end - start == 0:
            return nums[start]
        mid = (start + end) // 2
        left = self.findMinHelp(nums, start, mid)
        right = self.findMinHelp(nums, mid+1, end)
        return min(left, right)

if __name__ == "__main__":
    nums = [1,3,5]
    # nums = [2,2,2,0,1]
    result = Solution()
    print(result.findMin(nums))